### PR TITLE
[1/N][diffusion, rollout] feat: support HunyuanImage-3.0 Flow-GRPO training

### LINKED ISSUE #207: [RFC] Design and Implementation Plan for HunyuanImage 3.0 Support
### 1. Motivation
`HunyuanImage 3.0` is an 80B native multimodal model (13B active parameters) that unifies multimodal understanding and generation within an autoregressive (AR) and diffusion framework. It supports multiple generation modes including `t2t`, `i2t`, `t2i`, `it2i`, and a novel `t2ti` mode (think-recaption chain: AR chain-of-thought followed by diffusion generation). 

Integrating `HunyuanImage 3.0` into `verl-omni` is a critical step for scaling our RL training capabilities to state-of-the-art MoE-based unified multimodal models. It will allow us to perform end-to-end RL (e.g., GRPO for the AR reasoning phase and Flow-GRPO for the DiT phase) on a single unified architecture. 

### 2. Current Problem
Currently, `verl-omni` supports single-modality or decoupled architectures (like Qwen-Image for diffusion and Qwen3/Qwen-VL for AR). Introducing `HunyuanImage 3.0` presents new architectural and infrastructural challenges:
1. **Unified Two-Engine Pipeline (AR + DiT)**: The model is neither purely a diffusion model nor purely an AR model. It requires a unified trainer and rollout loop that can orchestrate both the AR engine (for CoT reasoning/captioning) and the DiT engine (for pixel generation) within the same RL step.
2. **Scale & MoE Architecture**: At 80B total parameters (MoE), the model exceeds the memory limits of single-GPU and potentially single-node setups without aggressive offloading. We must ensure FSDP/Megatron (for training) and `vllm-omni` (for rollout) can handle MoE routing and distributed parallelism efficiently.
3. **Sequential Modality Generation (`t2ti`)**: Supporting the `t2ti` mode requires custom rollout orchestration where the output tokens of the AR phase directly condition the diffusion phase before the final reward is calculated.

### 3. Proposed Design & Changes
Extending `verl-omni` to feature a **Unified Model Trainer and Rollout mechanism**.

1. **Unified Model Pipeline**:
   - Create a new pipeline directory (e.g., `verl_omni/pipelines/unified_model/`) that orchestrates both AR and DiT RL components.
   - For `t2ti` rollouts, the agent will query the `vllm-omni` AR engine to generate the CoT text, and immediately pass the generated text state into the `vllm-omni` DiT engine to generate the image. The joint trajectory will be evaluated by the reward function.

2. **MoE and Multi-Node Support**:
   - Leverage multi-node FSDP (or Megatron-LM) to distribute the 80B MoE model during the training phase.
   - On the `vllm-omni` rollout side, we will configure tensor parallelism (`TP=8`) to fit the replica within a single node if possible, or build upon the multi-node rollout design to support larger slices.

3. **Shared AR/DiT Plumping**:
   - Structure the AR step as a transition-level kernel (similar to `models/types/ar.py` in unirl) to cleanly separate the Markov transitions of the AR chain from the continuous time steps of the Flow-matching chain, while sharing the same underlying optimization loop.

4. **vLLM-Omni Integration**:
   - Extend `vllm-omni` to natively serve the `HunyuanImage 3.0` weight layout, handling both the LLM and the Diffusion transformer blocks efficiently during generation.

### 4. Dev Plan & Validation

**Implementation Plan:**
- [ ] **Model Definition**: Add `HunyuanImage 3.0` (Instruct) support to `verl-omni`'s model registry, handling its specific `trust_remote_code` or native loading paths.
- [ ] **Unified Pipeline**: Implement `train_unified_model.py` and configuration files (e.g., `examples/unified_model/hi3_vllmomni.yaml`) for end-to-end training.
- [ ] **Rollout Orchestration**: Update the `vllm-omni` worker to support the sequential `t2ti` rollout logic seamlessly.
- [ ] **Weight Sync**: Ensure the ZMQ IPC weight synchronization works correctly across AR and DiT blocks and handles MoE experts efficiently.

**Validation Plan:**
- **Modality Smoke Tests**: Validate `t2t`, `i2t`, `t2i`, and `t2ti` generation on a pristine `HunyuanImage-3.0-Instruct` checkpoint (single-process) to verify basic model compatibility.
- **End-to-End RL Loop**: Run a complete unified training step (FSDP-80B + AR/DiT vLLM-Omni engines -> Rollout -> PickScore Reward -> GRPO/Flow-GRPO -> Weight Sync). Verify that the rollout produces coherent images with valid rewards and that the training step computes sensible loss and gradient norms.
- **Scale Validation**: Test on a multi-node cluster (e.g., 2 nodes of 8xH800) to guarantee FSDP and vLLM-Omni scale properly without OOM issues or weight sync deadlocks.

### 5. Future TODOs
- **Expert Parallelism (EP)**: Investigate and integrate EP for both the FSDP trainer and `vllm-omni` to optimize the memory and compute of the MoE routing beyond standard Data Parallelism / Tensor Parallelism.
- **Broaden Unified Support**: Leverage this new `unified_model` architecture to eventually support other AR+DiT models (like Bagel) under the same framework.

### LINKED ISSUE #6768: [Feature]: Expose AR-stage rollout metadata (token ids / per-token log-probs / prompt token ids) via `DiffusionOutput.custom_output`
### 🚀 The feature, motivation and pitch

RL frameworks built on vllm-omni (e.g. unified AR+DiT training over HunyuanImage-3 https://github.com/verl-project/verl-omni/issues/207) need to teacher-force the AR recaption segment against the tokens actually sampled during rollout, and to use the sampled per-token log-probs as `old_log_probs`.

These values are already computed at inference time, then dropped — only `ar_generated_text` survives via `custom_output`, and re-tokenizing it is lossy (structure tokens like `<img_size_*>`) and cannot recover log-probs, so they cannot be re-derived client-side.

Would it be possible to expose, alongside `ar_generated_text` and through the same `custom_output` path:

- `ar_generated_token_ids` — from `output.cumulative_token_ids`
- `ar_generated_logprobs` — one per generated token, from the sampled token's entry in `output.logprobs`
- `ar_prompt_token_ids` — the AR request's prompt ids, so the teacher-forced prefix is guaranteed to match what the server actually consumed (robust to client-side tokenizer / chat-template drift)

(A generic per-request metadata dict channel would cover future fields, but naming that is maintainers' call.)

### Additional context

_No response_

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://vllm-omni.readthedocs.io), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by Xenoryn**
Close https://github.com/verl-project/verl-omni/issues/207

### What does this PR do?

Brings HunyuanImage-3.0-Instruct into verl-omni as a pipeline package. It is an 80B unified AR+DiT MoE transformer whose checkpoint is in transformers (non-diffusers) format, and whose rollout is a native two-stage vllm-omni pipeline (AR stage samples think/recaption text → DiT stage produces the image). This PR delivers four pieces:

1. **Training-side model definition** `HunyuanImage3ForTraining` — FSDP2-compatible with layer-level sharding, generation backbone ported from Tencent's official `modeling_hunyuan_image_3.py`;
2. **FlowGRPO training adapter** — the text condition at training time matches the rollout's sampled condition token-for-token (see below), CFG via batch doubling aligned with vllm-omni;
3. **Rollout adapter + agent loop** — both sides construct the same scheduler so the sigma schedule matches to the bit (prerequisite for an unbiased importance-sampling ratio); the agent loop builds the official AR prompt client-side through `prompt_utils` (the checkpoint ships no chat template);
4. **Two small shared-code changes** — the vllm-omni async server forwards AR sampling metadata through a whitelist; throughput metrics gain token counts. Both are backward compatible (details below).

**Scope boundary (stated up front):**
- This PR alone cannot run end-to-end training. The AR segment only samples during rollout and receives no gradient here; joint training lands in [2/N].
- End-to-end additionally requires: [the [1/N] multi-stage weight-sync ZMQ rank fix](https://github.com/verl-project/verl-omni/pull/483), two small upstream vllm-omni fixes [(#6768 sampling metadata](https://github.com/vllm-project/vllm-omni/issues/6768), [#6770](https://github.com/vllm-project/vllm-omni/issues/6770)), and the config / data-preprocessing scripts from the follow-up recipe PR.
- At the code level this PR depends on none of the above commits / upstream patches: it references nothing unmerged, degrades gracefully when metadata is absent, is self-contained for review, and does not affect existing models.

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here: ...
- [x] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `vllm_omni`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`, `diffusion`, `omni`, `tests`, `docker`
  - If this PR involves multiple modules, separate them with `,` like `[diffusion, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][diffusion, fsdp] feat: new rollout scheduler`

### Test

17 new CPU test cases, all passing:

```text
$ python -m pytest tests/pipelines/test_hunyuan_image3_flow_grpo_on_cpu.py -x -q
17 passed, 23 warnings in 0.29s
```

Coverage:

- `(HunyuanImage3ForCausalMM, flow_grpo)` registration on both the training and rollout sides, plus agent-loop registration;
- the local scheduler matches diffusers `FlowMatchEulerDiscreteScheduler` (same constructor kwargs as vllm-omni) on sigma/timesteps **to the bit**;
- CFG formula and sigma config (shift=3.0 exponential schedule);
- `prepare_model_inputs` positive/negative branches (each with its own `<timestep>` position and 2D RoPE spans);
- tiny-model forward returns a latent-space velocity; checkpoint `config.json` parsing; `from_pretrained` sharded-checkpoint round-trip;
- **deterministic MoE routing under ties** (upstream `torch.topk` is unstable for ties, so a recompute can flip a token's expert and corrupt gradients; fixed with a stable sort, locked by the test);
- AR text rmpad forward + varlen attention per-sample isolation equivalence (locks out cross-sample attention);
- fused DiT input rebuild (mocked vllm-omni tokenizer: cot normalization, system prompt, cfg_factor, and other rollout-alignment knobs) + fallback when AR fields are absent.

Note: server-side metadata forwarding and the real rollout trajectory are runtime behavior not covered by CPU tests; GPU smoke coverage is planned for a follow-up PR (TBD).

### API and Usage Example

Full training config / launch script / data preprocessing will follow in a later PR.

### Design & Code Changes

New package `verl_omni/pipelines/hunyuan_image3_flow_grpo/` (~2.4k lines).

Shared-file changes (no effect on other models by default, but flagged for reviewer attention):

- `vllm_omni_async_server.py`:
  1. Forwards `ar_prompt_token_ids / ar_generated_token_ids / ar_generated_text / ar_generated_logprobs` from `custom_output` through a whitelist (depends on upstream #6768; without the patch these fields are naturally absent and nothing errors);
  2. **in deploy-config mode, no longer passes the top-level `--tensor-parallel-size` to vllm-omni**. Reason: vllm-omni folds the top-level value into every stage's `parallel_config`, overriding the per-stage TP in the deploy config. No impact on existing in-repo users: bagel is single-GPU TP=1; the server-generated deploy config already writes TP into each stage;
  3. the `compilation_config` None-cleanup moves from AR-only to the shared path (a no-op on the vllm 0.24 dataclass path; avoids the explicit-None rejection on the 0.27 pydantic path).
- `diffusion_metric_utils.py`: token metrics (`perf/total_num_tokens`, `perf/tokens_per_sec`, `perf/ar_response_len_*`) are appended only when the batch carries `prompt_token_ids` / `ar_generated_token_ids`; output is byte-identical to before when they are absent.
- `pipelines/__init__.py`, `agent_loop/__init__.py`: register the new pipeline / agent loop (same pattern as the existing minimax and ltx2 ones).

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [x] Read the [Contribute Guide](https://github.com/verl-project/verl-omni/blob/main/CONTRIBUTING.md).
- [x] Apply [pre-commit checks](https://github.com/verl-project/verl-omni/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update the documentation.
- [x] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/verl-project/verl-omni/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...

### CODE REVIEW

**Comment by wtomin** [comment]
@Xenoryn Thanks for your contributions. Please resolve the conflicts. Afterwards, could you rebase to the lastest main branch, and rerun the training? Please provide the training example script, the document. And also supplement the training/validation rewards, as well as loss curves, in the PR body.

**Comment by NancyFyong** [comment]
Please resolve the conflicts so that we can review. Thank you!
